### PR TITLE
kata 2.0: add debug console service

### PR DESCRIPTION
**Pull request by liubin**
Add `kata-runtime exec` to enter guest OS through shell started by agent

This approach use `vsock` to connect shell running inside guest that started by agent. The good aspect is that we need not modify guest image or despite using what device that hypervisor support. Only `/bin/sh` or `/bin/bash` are necessary.

Fixes: #245

Signed-off-by: bin liu <bin@hyper.sh>

### CODE REVIEW

File: docs/Developer-Guide.md
Comment (bergwolf) [inline docs/Developer-Guide.md:498]: Can we make it internal so that users are not aware of which vsock port the debug console is using? For example:
1. add a new option `enable_debug_console`
2. when enabled, append `agent.debug_console agent.debug_console_vport=1026` to guest kernel command line automatically

Comment (liubin) [reply docs/Developer-Guide.md:498]: Yes, I think we can add `enable_debug_console` to `agent` section of Kata's TOML configuration file.

File: src/agent/src/main.rs
Comment (bergwolf) [inline src/agent/src/main.rs:457]: Would it loop forever and create tons of console threads?

Comment (liubin) [reply src/agent/src/main.rs:457]: No, there will only one shell process normally. In `run_debug_console_shell()` parent process will wait child process to exit and then enter the next loop. There will be only one shell process at one time.

**Comment by liubin** [comment]
CI failed due to https://github.com/kata-containers/tests/search?q=enable_debug&unscoped_q=enable_debug

File: src/runtime/cli/kata-exec.go
Comment (jodh-intel) [inline src/runtime/cli/kata-exec.go:31]: I think it would be clearer to specify this in kb.

File: src/runtime/cli/config-generated.go.in
Comment (jodh-intel) [inline src/runtime/cli/config-generated.go.in:39]: I don't think you need this change? This file is for code generated by the build, but this variable isn't generated by the build :smile:

File: src/runtime/cli/kata-exec.go
Comment (jodh-intel) [inline src/runtime/cli/kata-exec.go:46]: This can be simply:

```go
Name: "exec",
```

File: src/runtime/cli/kata-exec.go
Comment (jodh-intel) [inline src/runtime/cli/kata-exec.go:50]: Strictly, these should probably be prefixed with `kata-`:

```sh
$ ./kata-runtime --help | grep extensions
- Commands starting "kata-" and options starting "--kata-" are Kata Containers extensions.
```

File: src/runtime/cli/kata-exec.go
Comment (jodh-intel) [inline src/runtime/cli/kata-exec.go:68]: Can you define `"http://localhost:8090"` and `1026` (below) as global `const`s?

File: src/runtime/cli/kata-exec.go
Comment (jodh-intel) [inline src/runtime/cli/kata-exec.go:73]: `kernelParamDebugConsoleVPortValue`?

File: src/runtime/cli/kata-exec.go
Comment (jodh-intel) [inline src/runtime/cli/kata-exec.go:192]: This is rather cryptic.

File: docs/Developer-Guide.md
Comment (jodh-intel) [inline docs/Developer-Guide.md:489]: Rather than putting the details in brackets, could you just change the structure of the MD file to something like the following as that will be clearer when viewed in the table of contents:

```md
## Set up a debug console

    :

### Simple debug console setup

    :

### Traditional debug console setup
```

File: docs/Developer-Guide.md
Comment (jodh-intel) [inline docs/Developer-Guide.md:491]: I think we can simplify this to something like:

> Kata Containers 2.0 supports a shell simulated *console* for quick debug purpose. This approach uses VSOCK to 
> connect to the shell running inside the guest which the agent starts. This method only requires the guest image to
> contain either `/bin/sh` or `/bin/bash`.

File: docs/Developer-Guide.md
Comment (jodh-intel) [inline docs/Developer-Guide.md:495]: Enable  `debug_console_enabled` in the `configuration.toml` configuration file:

File: docs/Developer-Guide.md
Comment (jodh-intel) [inline docs/Developer-Guide.md:506]: > The `kata-runitime exec` command needs `kata-monitor` to ...

File: docs/Developer-Guide.md
Comment (jodh-intel) [inline docs/Developer-Guide.md:502]: s/`vsock`/VSOCK/g

File: src/agent/src/main.rs
Comment (jodh-intel) [inline src/agent/src/main.rs:441]: This is going to break agent shutdown (see https://github.com/kata-containers/kata-containers/pull/411, specifically the calls to `select!()` on https://github.com/kata-containers/kata-containers/pull/411/commits/4465190f49e01a0675e03eee8b6d52eb891ed022).

Comment (liubin) [reply src/agent/src/main.rs:441]: Thank you, @jodh-intel . That's mean we should hold this PR after #441 has been merged, and then adopt to the gracefully shutdown mechanism?

**Review by bergwolf — APPROVED** [review]
Thanks @liubin ! This is very helpful!

**Review by lifupan — APPROVED** [review]

**Pull request merged**
